lib/spack/spack/environment/environment.py

```diff
@@ -1249,16 +1249,23 @@ def develop(self, spec: Spec, path: str, clone: bool = False) -> bool:
             tty.msg("Configuring spec %s for development at path %s" % (spec, path))
 
         if clone:
-            # "steal" the source code via staging API
-            abspath = spack.util.path.canonicalize_path(path, default_wd=self.path)
-
-            # Stage, at the moment, requires a concrete Spec, since it needs the
-            # dag_hash for the stage dir name. Below though we ask for a stage
-            # to be created, to copy it afterwards somewhere else. It would be
+            # "steal" the source code via staging API. We ask for a stage
+            # to be created, then copy it afterwards somewhere else. It would be
             # better if we can create the `source_path` directly into its final
             # destination.
+            abspath = spack.util.path.canonicalize_path(path, default_wd=self.path)
             pkg_cls = spack.repo.path.get_pkg_class(spec.name)
-            pkg_cls(spec).stage.steal_source(abspath)
+            # We construct a package class ourselves, rather than asking for
+            # Spec.package, since Spec only allows this when it is concrete
+            package = pkg_cls(spec)
+            if isinstance(package.fetcher[0], spack.fetch_strategy.GitFetchStrategy):
+                package.fetcher[0].get_full_repo = True
+                # If we retrieved this version before and cached it, we may have
+                # done so without cloning the full git repo; likewise, any
+                # mirror might store an instance with truncated history.
+                package.stage[0].mirror_paths = []
```

Comment:
As an overall strategy I think it is fine to tweak the fetcher/stage options after the package has constructed them.

Some refactoring would be useful, for example replacing

```
package.stage[0].mirror_paths = []
```

with

```
package.stage[0].disable_mirrors()
```

(and likewise add a function like `retrieve_everything` to `FetchStrategy`). FWIW, I don't think it would be necessary (or cleaner) to pass parameters to `_make_fetcher` (or otherwise record this external desire to retrieve the full repository as an attribute on the package).

Does that seem reasonable to you, or do you think more refactoring would be necessary?

Reply:
Broadly speaking, in the long term I'd like to simplify `Fetcher` and `Stage` and their construction. Ideally, I'd also like to be able to specify properties when we assemble `package`, rather than tweaking internals after the fact in some function. This is clearly not a request for this PR.

> Some refactoring would be useful, for example ...

I agree with that if you want to do it here. Otherwise this PR LGTM

Reply:
I have done this for `Stage.disable_mirrors`, but it turns out that Git is the only fetcher that pulls a limited set of repo data to save space, so I did not add a `retrieve_everything` method to `FetchStrategy`.

> Broadly speaking, in the long term I'd like to simplify Fetcher and Stage and their construction. Ideally, I'd also like to be able to specify properties when we assemble package, rather than tweaking internals after the fact in some function. This is clearly not a request for this PR.

Fair enough, my only request is, as you proceed with a refactor, be open to the possibility of allowing users to change download settings after the download object has been created (i.e., in my opinion this aspect of the logic is not why fetching/staging should be refactored).